NF-545 Pass x-correlation-id, x-request-id and x-session-id to EIS/file transmission synchronous

crosslandwa opened the pull request:
Mandates that incoming amend/create requests specify an `X-Correlation-ID` header
 - pass this on as the `x-correlation-id` header to EIS
 - pass this on as the `conversation-id` header to File Transmission Synchronous
 
 Also passes x-request-id and x-session-id headers to EIS 
 
Note this is dependent on https://github.com/hmrc/national-duty-repayment-center-frontend/pull/405/files being merged/deployed first (as an `x-correlation-id` header is now required on the incoming request from the FE)

crosslandwa committed:
NF-545 Remove unused correlationId parameter

crosslandwa committed:
NF-545 Require incoming request to specify a correlation ID

Use this as the correlation ID for EIS create/amend requests
Use this as the conversation ID for EIS file transfer requests

crosslandwa committed:
NF-545 Add MDTP (tracing) headers to EIS requests

crosslandwa committed:
NF-545 Simplify FileTransferRequest model

Remove requestId field (not used in receiving service)
Make correlationId non-Optional (it is always supplied)

crosslandwa committed:
NF-545 Simplify generation of MDTPTracingHeaders

Also remove duplicate tracing headers on requests to internal stubs

crosslandwa committed:
NF-545 Accomodate FileTransferRequest changes in integration tests

SteveSugden commented on app/uk/gov/hmrc/nationaldutyrepaymentcenter/connectors/AmendCaseConnector.scala:
```diff
@@ -39,19 +39,23 @@ class AmendCaseConnector @Inject()(
         url,
         request,
-        eisApiHeaders(correlationId, config.eisEnvironment, config.eisAuthorizationToken)
+        EISApiHeaders(correlationId, config.eisEnvironment, config.eisAuthorizationToken) ++ MDTPTracingHeaders(hc)
```
Observation.  Not a fan of capitalised method names.  Looks like you are constructing two objects above.

crosslandwa replied:
Updated

SteveSugden commented on app/uk/gov/hmrc/nationaldutyrepaymentcenter/connectors/CreateCaseConnector.scala:
```diff
@@ -48,11 +48,15 @@ class CreateCaseConnector @Inject()(
-        hc.copy(authorization = None),
+        hc.copy(
+          authorization = None, // sent via EISApiHeaders
+          requestId = None, // sent via MDTPTracingHeaders (None here prevents sending duplicate header to internal stubs)
```
Just an observation, but I'm more confused about this comment than anything else.  Should we be hacking the code to improve the experience when using our stubs?

crosslandwa replied:
I was aiming for getting the behaviour (in terms of what headers are sent) on `QA -> EIS` and `local -> stubs` the same, to reduce having to think about differences in behaviour based on environment...

(...which is somewhat forced upon us by the underlying `HeaderCarrier` change whereby its _only_ used for "internal" requests...)

Do any of @SteveSugden , @lambertegbe1 , @njmoorewitchcat and @rorticus0502 have an opinion on which is preferred:
1. "expected" difference in what headers are sent when sending requests to stubs vs EIS
1. confusing comments/hacks in the code to enforce consistent behaviour when sending requests to stubs vs EIS


(if there's a consensus on 2, we can work on making the comments/code less confusing)

SteveSugden replied:
I'd vote for 2.  But just have the comment `// sent via MDTPTracingHeaders` which is similar to the line above.
Would that reduce the chance of someone removing those two lines in future thinking "its only for stubs, lets get rid")?

crosslandwa replied:
Updated this so that the `x-request-id` and `x-session-id` values in the header carrier are not manipulated - this is to avoid risk of regression in sending implicit auditing events

SteveSugden commented on app/uk/gov/hmrc/nationaldutyrepaymentcenter/models/FileTransferRequest.scala:
```diff
@@ -29,10 +29,7 @@ case class FileTransferRequest(
-                                              correlationId: Option[String] = None,
-                                              // private field, this value will be overwritten
-                                              // with x-request-id header value in the controller
-                                              requestId: Option[String] = None
```
Presume the `requestId` wasn't used here?

crosslandwa replied:
Correct

This class was copy/pasted from the File Transmission Synchronous service, which populates the `requestId` (a "private field") based on the incoming `x-request-id` header (overwriting whatever was de-serialised from the incoming payload)

https://github.com/hmrc/file-transmission-synchronous/blob/master/app/uk/gov/hmrc/traderservices/controllers/FileTransferController.scala#L64

SteveSugden reviewed (dismissed):
LGTM - just a couple of observations

crosslandwa committed:
NF-545 Use non-capitalised method names

crosslandwa committed:
NF-545 Do not manipulate tracing headers in header carrier

Removing these risks breaking implicit auditing. Accept that this means we'll see duplicate x-request-id and x-session-id headers when sending (EIS) requests to our (internal) stub

crosslandwa committed:
NF-545 Generate a valid AcknowledgmentReference in EIS payload.

Also make x-correlation-id header optional (for backwards compatibility with current FE)

crosslandwa committed:
Formatting

SteveSugden commented on app/uk/gov/hmrc/nationaldutyrepaymentcenter/controllers/WithCorrelationId.scala:
```diff
@@ -0,0 +1,41 @@
+import scala.concurrent.Future
+
+trait WithCorrelationId { self: Results =>
+  val missingXCorrelationIdResponse: Result = BadRequest(
```
So this val is currently unused now?   Don't remove, just checking my understanding

crosslandwa replied:
Unused - should have removed it... though it will be used again (instead of the UUIDGenerator) in a followup change
